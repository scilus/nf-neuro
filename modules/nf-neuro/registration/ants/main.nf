
process REGISTRATION_ANTS {
    tag "$meta.id"
    label 'process_medium'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
    tuple val(meta), path(fixedimage), path(movingimage), path(mask) //** optional, input = [] **//

    output:
    tuple val(meta), path("*_warped.nii.gz")                                  , emit: image
    tuple val(meta), path("*__output{0Warp.nii.gz,1GenericAffine.mat}")       , emit: transfo_image
    tuple val(meta), path("*__output{0,1}Inverse{Warp.nii.gz,Affine.mat}")    , emit: transfo_trk
    path "versions.yml"                                                       , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def ants = task.ext.quick ? "antsRegistrationSyNQuick.sh " :  "antsRegistrationSyN.sh "
    def dimension = task.ext.dimension ? "-d " + task.ext.dimension : "-d 3"
    def transform = task.ext.transform ? task.ext.transform : "s"
    def seed = task.ext.random_seed ? " -e " + task.ext.random_seed : "-e 1234"

    if ( task.ext.threads ) args += "-n " + task.ext.threads
    if ( task.ext.initial_transform ) args += " -i " + task.ext.initial_transform
    if ( task.ext.histogram_bins ) args += " -r " + task.ext.histogram_bins
    if ( task.ext.spline_distance ) args += " -s " + task.ext.spline_distance
    if ( task.ext.gradient_step ) args += " -g " + task.ext.gradient_step
    if ( task.ext.mask ) args += " -x $mask"
    if ( task.ext.type ) args += " -p " + task.ext.type
    if ( task.ext.histogram_matching ) args += " -j " + task.ext.histogram_matching
    if ( task.ext.repro_mode ) args += " -y " + task.ext.repro_mode
    if ( task.ext.collapse_output ) args += " -z " + task.ext.collapse_output

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=$task.cpus
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    $ants $dimension -f $fixedimage -m $movingimage -o output -t $transform $args $seed

    mv outputWarped.nii.gz ${prefix}__warped.nii.gz
    mv output0GenericAffine.mat ${prefix}__output1GenericAffine.mat

    if [ $transform != "t" ] && [ $transform != "r" ] && [ $transform != "a" ];
    then
        mv output1InverseWarp.nii.gz ${prefix}__output1InverseWarp.nii.gz
        mv output1Warp.nii.gz ${prefix}__output0Warp.nii.gz
    fi

    antsApplyTransforms -d 3 -i $fixedimage -r $movingimage -o Linear[output.mat]\
                        -t [${prefix}__output1GenericAffine.mat,1]

    mv output.mat ${prefix}__output0InverseAffine.mat


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ants: antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9]+\\.[0-9]+\\.[0-9]+).*/\\1/'
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    antsRegistrationSyNQuick.sh -h
    antsApplyTransforms -h

    touch ${prefix}__t1_warped.nii.gz
    touch ${prefix}__output1GenericAffine.mat
    touch ${prefix}__output0InverseAffine.mat
    touch ${prefix}__output1InverseWarp.nii.gz
    touch ${prefix}__output0Warp.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ants: antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9]+\\.[0-9]+\\.[0-9]+).*/\\1/'
    END_VERSIONS
    """
}
