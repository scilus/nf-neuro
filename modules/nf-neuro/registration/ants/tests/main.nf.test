nextflow_process {

    name "Test Process REGISTRATION_ANTS"
    script "../main.nf"
    process "REGISTRATION_ANTS"

    tag "modules"
    tag "modules_nfcore"
    tag "registration"
    tag "registration/ants"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "T1w.zip", "b0.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("registration - ants") {
        config "./nextflow.config"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        T1w: it.simpleName == "T1w"
                        b0: it.simpleName == "b0"
                    }
                ch_T1w = ch_split_test_data.T1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                ch_b0 = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]
                }
                input[0] = ch_b0
                    .join(ch_T1w)
                    .map{ it + [[]] }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("registration - ants - quick") {
        config "./nextflow_quick.config"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        T1w: it.simpleName == "T1w"
                        b0: it.simpleName == "b0"
                    }
                ch_T1w = ch_split_test_data.T1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                ch_b0 = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]
                }
                input[0] = ch_b0
                    .join(ch_T1w)
                    .map{ it + [[]] }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("registration - ants - options") {
        config "./nextflow_options.config"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        T1w: it.simpleName == "T1w"
                        b0: it.simpleName == "b0"
                    }
                ch_T1w = ch_split_test_data.T1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz"),
                        file("\${test_data_directory}/T1w_mask.nii.gz")
                    ]
                }
                ch_b0 = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]
                }
                input[0] = ch_b0
                    .join(ch_T1w)
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
