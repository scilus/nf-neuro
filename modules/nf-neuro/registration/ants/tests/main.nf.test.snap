{
    "registration - ants": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,4382217c04988eb01ed8f99916d7631c"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test__output0Warp.nii.gz:md5,6fee747df2c30bb48cf2b6897b5d3e17",
                            "test__output1GenericAffine.mat:md5,8e987eb08cc568478c37ae79a8824f8b"
                        ]
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test__output0InverseAffine.mat:md5,d342d5f1c35a53cadcdf2d49a25f5ce8",
                            "test__output1InverseWarp.nii.gz:md5,1608837de13478dff81fe51ac1720d28"
                        ]
                    ]
                ],
                "3": [
                    "versions.yml:md5,825e9d4ea9e26675d7e236f243d73114"
                ],
                "image": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,4382217c04988eb01ed8f99916d7631c"
                    ]
                ],
                "transfo_image": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test__output0Warp.nii.gz:md5,6fee747df2c30bb48cf2b6897b5d3e17",
                            "test__output1GenericAffine.mat:md5,8e987eb08cc568478c37ae79a8824f8b"
                        ]
                    ]
                ],
                "transfo_trk": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test__output0InverseAffine.mat:md5,d342d5f1c35a53cadcdf2d49a25f5ce8",
                            "test__output1InverseWarp.nii.gz:md5,1608837de13478dff81fe51ac1720d28"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,825e9d4ea9e26675d7e236f243d73114"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-25T17:37:35.205440745"
    },
    "registration - ants - quick": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,81358dd2f3223f567f4101e61490ae0f"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test__output0Warp.nii.gz:md5,24e6f53eaa3e4476bb627c374bc809ca",
                            "test__output1GenericAffine.mat:md5,48c8479a403e0c0ec29c69d37bb4c91a"
                        ]
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test__output0InverseAffine.mat:md5,25d0b8a83a922f2e4e89c6a86258af15",
                            "test__output1InverseWarp.nii.gz:md5,24e6f53eaa3e4476bb627c374bc809ca"
                        ]
                    ]
                ],
                "3": [
                    "versions.yml:md5,825e9d4ea9e26675d7e236f243d73114"
                ],
                "image": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,81358dd2f3223f567f4101e61490ae0f"
                    ]
                ],
                "transfo_image": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test__output0Warp.nii.gz:md5,24e6f53eaa3e4476bb627c374bc809ca",
                            "test__output1GenericAffine.mat:md5,48c8479a403e0c0ec29c69d37bb4c91a"
                        ]
                    ]
                ],
                "transfo_trk": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test__output0InverseAffine.mat:md5,25d0b8a83a922f2e4e89c6a86258af15",
                            "test__output1InverseWarp.nii.gz:md5,24e6f53eaa3e4476bb627c374bc809ca"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,825e9d4ea9e26675d7e236f243d73114"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-25T17:14:19.681843788"
    },
    "registration - ants - options": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,7cb6839d4af18d9d37a1bd9d07c8c57b"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output1GenericAffine.mat:md5,c3a48d19c815206de2b140f505969317"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output0InverseAffine.mat:md5,f473c1e548be1540cbf757202ffd5984"
                    ]
                ],
                "3": [
                    "versions.yml:md5,825e9d4ea9e26675d7e236f243d73114"
                ],
                "image": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,7cb6839d4af18d9d37a1bd9d07c8c57b"
                    ]
                ],
                "transfo_image": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output1GenericAffine.mat:md5,c3a48d19c815206de2b140f505969317"
                    ]
                ],
                "transfo_trk": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output0InverseAffine.mat:md5,f473c1e548be1540cbf757202ffd5984"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,825e9d4ea9e26675d7e236f243d73114"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-25T17:14:25.782091391"
    }
}